File: javascript/hasPathBFS.js
const hasPath = (graph, src, dst) => {
  // code here
  const queue = [src];
  while (queue.length > 0) {
    const current = queue.shift();
    if (current === dst) return true;
    for (let node of graph[current]) {
      //   if (node === dst) {
      //     return true;
      //   }
      queue.push(node);
    }
  }
  return false;
};

const graph = {
  a: ["b", "c"],
  b: ["d"],
  c: ["e"],
  d: ["f"],
  e: [],
  f: [],
  k: [],
};

console.log(hasPath(graph, "a", "f"));
console.log(hasPath(graph, "b", "k"));

File: python/has_path_bfs.py
def has_path(graph, src, dst):
    queue = [src]
    while len(queue) > 0:
        current = queue.pop()
        for node in graph[current]:
            if node == dst:
                return True
            queue = [node] + queue
    return False

graph = {
    "f" : ["g", "i"],
    "i" : ["g", "k"],
    "k" : [],
    "g": ["h"],
    "h" : []
}

print(has_path(graph, "f", "h"))
print(has_path(graph, "f", "j"))